packages/devtools_app/lib/src/shared/preferences.dart

```diff
@@ -114,25 +131,54 @@ class InspectorPreferencesController extends DisposableController
     });
   }
 
-  void _initCustomPubRootListeners() {
-    // TODO(CoderDake): add _customPubRootDirectories listener back when
-    // finalizing https://github.com/flutter/devtools/issues/3941
-    /*
-    addAutoDisposeListener(_customPubRootDirectories, () {
-      storage.setValue(
-        _customPubRootDirectoriesStorageId,
-        jsonEncode(_customPubRootDirectories.value),
-      );
-    });
-    */
+  Future<void> _initCustomPubRootDirectories() async {
+    autoDisposeStreamSubscription(
+      serviceManager.onConnectionAvailable
+          .listen(_handleConnectionToNewService),
+    );
+    autoDisposeStreamSubscription(
+      serviceManager.onConnectionClosed.listen(_handleConnectionClosed),
+    );
     addAutoDisposeListener(_busyCounter, () {
       _customPubRootDirectoriesAreBusy.value = _busyCounter.value != 0;
     });
   }
 
+  void _handleConnectionClosed(dynamic _) async {
+    _mainScriptDir = null;
+    _customPubRootDirectories.clear();
+  }
+
+  Future<void> _handleConnectionToNewService(VmServiceWrapper wrapper) async {
+    await _updateMainScriptRef();
+
+    _customPubRootDirectories.clear();
+
+    await loadCustomPubRootDirectories();
+
+    // Only setup the isolate listener the first time we connect to a service
+    if (isolateListener == null) {
+      isolateListener = () {
+        preferences.inspector.loadCustomPubRootDirectories();
+      };
+      addAutoDisposeListener(
+        serviceManager.isolateManager.mainIsolate,
+        isolateListener,
+      );
+    }
+  }
+
+  void _persistCustomPubRootDirectoriesToStorage() {
+    storage.setValue(
+      _customPubRootStorageId(),
+      jsonEncode(_customPubRootDirectories.value),
+    );
+  }
+
   Future<void> addPubRootDirectories(
     List<String> pubRootDirectories,
   ) async {
+    if (!serviceManager.hasConnection) return;
```

Comment:
It might be worth adding an error state to the pubrootdirectories in the future. and set this instead of just shortcircuiting